sklearn/utils/_random.pyx

```diff
@@ -310,3 +312,31 @@ cpdef sample_without_replacement(np.int_t n_population,
     else:
         raise ValueError('Expected a method name in %s, got %s. '
                          % (all_methods, method))
+
+
+
+# rand_r replacement using a 32bit XorShift generator
+# See http://www.jstatsoft.org/v08/i14/paper for details
+cdef inline UINT32_t our_rand_r(UINT32_t* seed) nogil:
```

Comment:
This won't be inlined if it's not in a header file. That is why it was defined repeatedly. Can we put it in a pxd? Or at least check the effect of the current change on runtime benchmarks for each affected use?